linkerd/tests/test_linkerd.py

```diff
@@ -546,40 +45,21 @@ def test_linkerd_v2(aggregator):
-    aggregator.assert_metric('linkerd.route.actual_response_latency.count', metric_type=aggregator.GAUGE)
-    aggregator.assert_metric('linkerd.route.actual_response_latency.sum', metric_type=aggregator.GAUGE)
-    aggregator.assert_metric('linkerd.route.actual_response_total', metric_type=aggregator.MONOTONIC_COUNT)
-    aggregator.assert_metric('linkerd.tcp.open_total', metric_type=aggregator.MONOTONIC_COUNT)
-    aggregator.assert_metric('linkerd.tcp.open_connections', metric_type=aggregator.GAUGE)
-    aggregator.assert_metric('linkerd.tcp.read_bytes_total', metric_type=aggregator.MONOTONIC_COUNT)
-    aggregator.assert_metric('linkerd.tcp.write_bytes_total', metric_type=aggregator.MONOTONIC_COUNT)
-    aggregator.assert_metric('linkerd.tcp.close_total', metric_type=aggregator.MONOTONIC_COUNT)
-    aggregator.assert_metric('linkerd.tcp.connection_duration.count', metric_type=aggregator.GAUGE)
-    aggregator.assert_metric('linkerd.tcp.connection_duration.sum', metric_type=aggregator.GAUGE)
-    aggregator.assert_metric('linkerd.control.request_total', metric_type=aggregator.MONOTONIC_COUNT)
-    aggregator.assert_metric('linkerd.control.response_latency.count', metric_type=aggregator.GAUGE)
-    aggregator.assert_metric('linkerd.control.response_latency.sum', metric_type=aggregator.GAUGE)
-    aggregator.assert_metric('linkerd.control.response_total', metric_type=aggregator.MONOTONIC_COUNT)
-    aggregator.assert_metric('linkerd.process.start_time', metric_type=aggregator.GAUGE)
-    aggregator.assert_metric('linkerd.process.cpu_seconds_total', metric_type=aggregator.MONOTONIC_COUNT)
-    aggregator.assert_metric('linkerd.process.open_fds', metric_type=aggregator.GAUGE)
-    aggregator.assert_metric('linkerd.process.max_fds', metric_type=aggregator.GAUGE)
-    aggregator.assert_metric('linkerd.process.virtual_memory', metric_type=aggregator.GAUGE)
-    aggregator.assert_metric('linkerd.process.resident_memory', metric_type=aggregator.GAUGE)
-
-    aggregator.assert_metric('linkerd.prometheus.health', metric_type=aggregator.GAUGE)
+    for metric_name, metric_type in EXPECTED_METRICS_V2.items():
+        aggregator.assert_metric(metric_name, metric_type=metric_type)
 
     aggregator.assert_all_metrics_covered()
 
     aggregator.assert_service_check(
-        'linkerd.prometheus.health', status=check.OK, tags=['endpoint:http://fake.tld/prometheus'], count=1
+        'linkerd.prometheus.health', status=LinkerdCheck.OK, tags=['endpoint:http://fake.tld/prometheus'], count=1
     )
+
+
+@pytest.mark.e2e
+def test_e2e(dd_agent_check):
+    aggregator = dd_agent_check(rate=True)
+    for metric_name, metric_type in EXPECTED_METRICS_V2_E2E.items():
+        aggregator.assert_metric(metric_name, metric_type=metric_type)
+    aggregator.assert_all_metrics_covered()
+
+    aggregator.assert_service_check('linkerd.prometheus.health', status=LinkerdCheck.OK, count=2)
```

Comment:
I'm getting this when running the tests (tried several times, same result).

Got this the first time:
```
E       AssertionError: Needed at least 1 candidates for 'linkerd.route.actual_response_latency.sum', got 0
E         Expected:
E                 MetricStub(name='linkerd.route.actual_response_latency.sum', type=0, value=None, tags=None, hostname=None)
E         Similar submitted:
E         Score   Most similar
E         0.59    MetricStub(name=u'linkerd.prometheus.health', type=0, value=1, tags=[u'endpoint:http://172.20.10.3:58720/metrics'], hostname=u'docker-desktop')
E         0.59    MetricStub(name=u'linkerd.prometheus.health', type=0, value=1, tags=[u'endpoint:http://172.20.10.3:58720/metrics'], hostname=u'docker-desktop')
E       assert False
```

But got this the second time, not sure why:

```
E       AssertionError: Needed at least 1 candidates for 'linkerd.route.actual_request_total', got 0
E         Expected:
E                 MetricStub(name='linkerd.route.actual_request_total', type=2, value=None, tags=None, hostname=None)
E         Similar submitted:
E         Score   Most similar
E         0.38    MetricStub(name=u'linkerd.prometheus.health', type=0, value=1, tags=[u'endpoint:http://172.20.10.3:59315/metrics'], hostname=u'docker-desktop')
E         0.38    MetricStub(name=u'linkerd.prometheus.health', type=0, value=1, tags=[u'endpoint:http://172.20.10.3:59315/metrics'], hostname=u'docker-desktop')
E       assert False
```

Full trace:

```
=========================================================================== test session starts ===========================================================================
platform darwin -- Python 2.7.16, pytest-4.6.5, py-1.8.0, pluggy-0.12.0 -- /Users/alexandre.yang/repos/dd/integrations-core/linkerd/.tox/py27/bin/python
cachedir: .tox/py27/.pytest_cache
benchmark: 3.2.2 (defaults: timer=time.time disable_gc=False min_rounds=5 min_time=0.000005 max_time=1.0 calibration_precision=10 warmup=False warmup_iterations=100000)
rootdir: /Users/alexandre.yang/repos/dd/integrations-core/linkerd
plugins: benchmark-3.2.2, cov-2.7.1, mock-1.10.4, datadog-checks-dev-0.31.1
collected 3 items / 2 deselected / 1 selected

tests/test_linkerd.py::test_e2e FAILED                                                                                                                              [100%]

================================================================================ FAILURES =================================================================================
________________________________________________________________________________ test_e2e _________________________________________________________________________________

dd_agent_check = <function run_check at 0x10b511aa0>

    @pytest.mark.e2e
    def test_e2e(dd_agent_check):
        aggregator = dd_agent_check(rate=True)
        for metric_name, metric_type in EXPECTED_METRICS_V2_E2E.items():
>           aggregator.assert_metric(metric_name, metric_type=metric_type)

tests/test_linkerd.py:62:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../datadog_checks_base/datadog_checks/base/stubs/aggregator.py:193: in assert_metric
    self._assert(condition, msg=msg, expected_stub=expected_metric, submitted_elements=self._metrics)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

condition = False, msg = "Needed at least 1 candidates for 'linkerd.route.actual_response_latency.sum', got 0"
expected_stub = MetricStub(name='linkerd.route.actual_response_latency.sum', type=0, value=None, tags=None, hostname=None)
submitted_elements = defaultdict(<type 'list'>, {u'linkerd.prometheus.health': [MetricStub(name=u'l...=[u'endpoint:http://172.20.10.3:58720/metrics'], hostname=u'docker-desktop')]})

    @staticmethod
    def _assert(condition, msg, expected_stub, submitted_elements):
        new_msg = msg
        if not condition:  # It's costly to build the message with similar metrics, so it's built only on failure.
            new_msg = "{}\n{}".format(msg, build_similar_elements_msg(expected_stub, submitted_elements))
>       assert condition, new_msg
E       AssertionError: Needed at least 1 candidates for 'linkerd.route.actual_response_latency.sum', got 0
E         Expected:
E                 MetricStub(name='linkerd.route.actual_response_latency.sum', type=0, value=None, tags=None, hostname=None)
E         Similar submitted:
E         Score   Most similar
E         0.59    MetricStub(name=u'linkerd.prometheus.health', type=0, value=1, tags=[u'endpoint:http://172.20.10.3:58720/metrics'], hostname=u'docker-desktop')
E         0.59    MetricStub(name=u'linkerd.prometheus.health', type=0, value=1, tags=[u'endpoint:http://172.20.10.3:58720/metrics'], hostname=u'docker-desktop')
E       assert False

../datadog_checks_base/datadog_checks/base/stubs/aggregator.py:235: AssertionError
================================================================= 1 failed, 2 deselected in 9.22 seconds ==================================================================
ERROR: InvocationError for command /Users/alexandre.yang/repos/dd/integrations-core/linkerd/.tox/py27/bin/pytest -v (exited with code 1)
_________________________________________________________________________________ summary _________________________________________________________________________________
ERROR:   py27: commands failed

```